### PR TITLE
Add support for allocating pages for code that can be marked read/execute

### PR DESCRIPTION
**Pull request by Alan-Jowett**
Add support for allocating pages for code that can be marked read/execute.

The first part of the fix for issue 217.

Use memory manager APIs to allocate physical pages and create a virtual address mapping for these pages.
Expose API to allocate pages
Expose API to modify the page protections (set them to either read-only, read & write, or read & execute). 
Expose API to release pages.
Expose API to get the base address of the pages.

To resolve issue 217, the following still needs to be done:
1) Add support to epoch code to invoke a function with context on epoch end (to release physical pages).
2) Switch ebpf_program_t to allocate pages from physical memory and mark them as read & execute after populating them.

### CODE REVIEW

**Review by poornagmsft — APPROVED** [review]

File: libs/platform/ebpf_platform.h [hunk 1]
```diff
@@ -124,6 +124,59 @@ extern "C"
...
+        EBPF_PAGE_PROTECT_READ_EXECUTE,
+    } ebpf_page_protection_t;
+
+    typedef struct _ebpf_memory_descriptor ebpf_memory_descriptor_t;
+
+    /**
+     * @brief Allocate pages from physical memory and create a mapping into the
+     * system address space.
+     *
+     * @param length Size of memory to allocate (internally this gets rounded
```
Comment (dthaler) [inline libs/platform/ebpf_platform.h:140]:
```suggestion
     * @param[in] length Size of memory to allocate (internally this gets rounded
```
Similarly please add [in,out] annotation throughout this file.

File: libs/platform/ebpf_platform.h [hunk 2]
```diff
@@ -124,6 +124,59 @@ extern "C"
...
+     * allocated pages.
+     * @param protection The new page protection to apply.
+     * @retval EBPF_SUCCESS The operation was successful.
+     * @retval EBPF_INVALID_ARGUMENT An invalid argument was supplied.
+     */
+    ebpf_result_t
+    ebpf_protect_memory(ebpf_memory_descriptor_t* memory_descriptor, ebpf_page_protection_t protection);
+
+    /**
+     * @brief Given a ebpf_memory_descriptor_t allocate via ebpf_map_memory
```
Comment (dthaler) [inline libs/platform/ebpf_platform.h:170]:
```suggestion
     * @brief Given an ebpf_memory_descriptor_t allocated via ebpf_map_memory,
```

File: libs/platform/ebpf_platform.h [hunk 3]
```diff
@@ -124,6 +124,59 @@ extern "C"
...
+     * obtain the base virtual address.
+     *
+     * @param memory_descriptor Pointer to ebpf_memory_descriptor_t describing
+     * allocated pages.
+     * @return void* Base virtual address of pages that have been allocated.
```
Comment (dthaler) [inline libs/platform/ebpf_platform.h:175]:
```suggestion
     * @return Base virtual address of pages that have been allocated.
```
Don't put the return type in the doxygen

File: libs/platform/user/ebpf_platform_user.cpp [hunk 4]
```diff
@@ -131,6 +131,72 @@ ebpf_free(void* memory)
...
+ebpf_unmap_memory(ebpf_memory_descriptor_t* memory_descriptor)
+{
+    if (memory_descriptor) {
+        VirtualFree(memory_descriptor->base, memory_descriptor->length, MEM_RELEASE);
+        free(memory_descriptor);
+    }
+}
+
+ebpf_result_t
+ebpf_protect_memory(ebpf_memory_descriptor_t* memory_descriptor, ebpf_page_protection_t protection)
```
Comment (dthaler) [inline libs/platform/user/ebpf_platform_user.cpp:169]: can memory_descriptor be const?

File: libs/platform/user/ebpf_platform_user.cpp [hunk 5]
```diff
@@ -131,6 +131,72 @@ ebpf_free(void* memory)
...
+    if (!VirtualProtect(
+            memory_descriptor->base, memory_descriptor->length, mm_protection_state, &old_mm_protection_state)) {
+        return EBPF_INVALID_ARGUMENT;
+    }
+
+    return EBPF_SUCCESS;
+}
+
+void*
+ebpf_memory_descriptor_get_base_address(ebpf_memory_descriptor_t* memory_descriptor)
```
Comment (dthaler) [inline libs/platform/user/ebpf_platform_user.cpp:196]:
```suggestion
ebpf_memory_descriptor_get_base_address(const ebpf_memory_descriptor_t* memory_descriptor)
```

File: libs/platform/ebpf_platform.h [hunk 6]
```diff
@@ -124,6 +124,59 @@ extern "C"
...
+    /**
+     * @brief Given an ebpf_memory_descriptor_t allocated via ebpf_map_memory
+     * obtain the base virtual address.
+     *
+     * @param[in] memory_descriptor Pointer to an ebpf_memory_descriptor_t
+     * describing allocated pages.
+     * @return Base virtual address of pages that have been allocated.
+     */
+    void*
+    ebpf_memory_descriptor_get_base_address(ebpf_memory_descriptor_t* memory_descriptor);
```
Comment (dthaler) [inline libs/platform/ebpf_platform.h:178]:
```suggestion
    ebpf_memory_descriptor_get_base_address(const ebpf_memory_descriptor_t* memory_descriptor);
```

File: libs/platform/kernel/ebpf_platform_kernel.c [hunk 7]
```diff
@@ -39,6 +44,65 @@ ebpf_free(void* memory)
...
+
+    status = MmProtectMdlSystemAddress(&memory_descriptor->memory_descriptor_list, mm_protection_state);
+    if (!NT_SUCCESS(status))
+        return EBPF_INVALID_ARGUMENT;
+
+    return EBPF_SUCCESS;
+}
+
+void*
+ebpf_memory_descriptor_get_base_address(ebpf_memory_descriptor_t* memory_descriptor)
```
Comment (dthaler) [inline libs/platform/kernel/ebpf_platform_kernel.c:101]:
```suggestion
ebpf_memory_descriptor_get_base_address(const ebpf_memory_descriptor_t* memory_descriptor)
```

**Pull request merged**